updog and migrator: signed migrations

webern opened the pull request:
## Issue number:

Closes #91
Closes #905 

## Description of changes:

Implements all changes necessary for migrator to load a locally cached tuf repo and obtain migrations from there. See #905 for a detailed explanation of the aigned migrations design.

Additionally, we decided to maintain compatibility so that instances can be upgraded into signed migrations. To facilitate this, in the current version, `upog` writes both signed and unsigned migrations. `migrator` checks the from version. If it's less than `0.3.5`, migrator executes unsigned migrations, otherwise it executes signed migrations.

In a future version, we can create a breaking change in which unsigned migrations are no-longer read. At that point, versions less than 0.3.5 must pass through a version that supports both signed and unsigned migrations (such as this one) in order to upgrade into signed migrations.

There is one behavioral change about signed migrations. Previously we used a `sort` of the migration names to provide stability to their order of execution (when more than one migration existed for a single version pair). With signed migrations, this is no longer necessary and the migrations will run in the order that they are listed in the manifest.

## Testing done:

The testing procedure is extensive. Note: a trace was added so that `migrator` announces in the system journal whether it is running signed or unsigned migrations. Also note, v0.3.1 was used as a starting point because a migration exists for v0.3.2 (which changes the host container version). Thus upgrading from v0.3.1 presents an actual unsigned migration.

### Test Setup
  
  * Create an AMI of v0.3.1, but with my non-production `root.json`.
  * Create matching image binaries for v0.3.1 with my non-production `root.json` and add them to a TUF repo (replacing the prod images).
  * Create a version of the code in this PR, naming the version 0.99.0, create an AMI of this and add the update binaries to the TUF repo/manifest.
  * Create a version of the code in this PR, along with a migration that adds a `foo` setting and sets its value to `bar`. Call this v0.99.1 and add it to the TUF repo/manifest.

### Test Execution

Starting from the v0.3.1 AMI, perform the following sequence of upgrades and downgrades.

  * Start at v0.3.1 via AMI
  * Upgrade v0.3.1 -> v0.99.0
  * Upgrade v0.99.0 -> v0.99.1
  * Downgrade v0.99.1 -> v0.99.0
  * Downgrade v0.99.0 -> v0.3.1

At each step along the way.

  * Ensure that Kubernetes is working by running a busybox pod.
  * Copy the system journal and Bottlerocket API settings to local storage.

At the end of the cycle, diff the system journals to observe that migrator annouced:

  * Upgrade v0.3.1 -> v0.99.0 'running unsigned migrations'
  * Upgrade v0.99.0 -> v0.99.1 'running signed migrations'
  * Downgrade v0.99.1 -> v0.99.0 'running signed migrations'
  * Downgrade v0.99.0 -> v0.3.1: older version of migrator made no announcement

Diff the API settings to observe the following setting changes:

  * Upgrade v0.3.1 -> v0.99.0: host container changed from v0.4.0 to v0.5.0
  * Upgrade v0.99.0 -> v0.99.1: "foo": "bar" was added
  * Downgrade v0.99.1 -> v0.99.0: "foo": "bar" was removed
  * Downgrade v0.99.0 -> v0.3.1: host container changed from v0.5.0 to v0.4.0

### Additional Testing

Additional testing has been performed with v0.99.0 AMI -> v0.99.1 -> v0.99.0, though this has not been repeated at every testing cycle.

## TODO

  * [x] Use `pentacle`
  * [x] Use `tough::RepoEditor` to create the `migrator` unit test repo on the fly.  ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  * [x] Use a single migration script for testing instead of two different migration scripts, and compress the migration script on the fly instead of checking in a compressed version of it. ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  * [x] Set repo expiration to 1970-01-01 to make it obvious that it is expired. Also make a note about this in the test. ([pushed](https://github.com/bottlerocket-os/bottlerocket/compare/f481a2dba97732110a08896f8ad009a2c49284ea..44434413686ee190eeba8a65cef0a1f315c027fe))
  
## Terms of contribution:

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

webern committed:
updog and migrator: signed migrations

webern commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +494,204 @@ where
+        }
+    }
+
+    /// This test ensures that migrations run when migrating from an older to a newer version.
```
```suggestion
    /// This test ensures that migrations run when migrating from a newer to an older version.
```

webern commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +494,204 @@ where
+                "Expected the migration 'x-first-migration.sh' to run second and write \
+            a message containing 'x-first-migration: --backward' to the output file. Instead \
+            found '{}'",
+                first_line
```
```suggestion
                second_line
```

webern commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -431,18 +415,34 @@ pub fn load_manifest<T: tough::Transport>(repository: &tough::Repository<T>) ->
+
+#[test]
+fn test_migrations_backward() {
+    // The same manifest as `test_migrations` but this time we will migrate backward.
```
```suggestion
    // The same manifest as `test_migrations_forward` but this time we will migrate backward.
```

webern commented on sources/api/migration/migrator/Cargo.toml:
```diff
@@ -18,10 +18,16 @@ semver = "0.9"
 snafu = "0.6"
 update_metadata = { path = "../../../updater/update_metadata" }
+tough = { version = "0.5.0" }
+url = "2.1.0"
```
```suggestion
```

All this is doing is adding `file://` in front of a path. Factor it out.

tjkirch commented on sources/api/migration/migrator/tests/data/repository/targets/e823375105cd38d5e0476a4e277bc61eef3c15bd0d8831e2f78bc9394b4f1335.manifest.json:
```diff
@@ -0,0 +1,12 @@
+  "migrations": {
+    "(0.3.1, 0.3.2)": [
+      "migrate_v0.3.2_admin-container-v0-5-0.lz4"
+    ],
```
Do we need this entry?

tjkirch commented on sources/api/migration/migrator/tests/data/repository/targets/e823375105cd38d5e0476a4e277bc61eef3c15bd0d8831e2f78bc9394b4f1335.manifest.json:
```diff
@@ -0,0 +1,12 @@
+    ],
+    "(0.99.0, 0.99.1)": [
+      "x-first-migration.lz4",
+      "a-second-migration.lz4"
```
Can we include the sources for these in the repo/PR?  A little creepy to have to trust executable binaries to run testing.

tjkirch commented on sources/api/migration/migrator/tests/data/repository/metadata/timestamp.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "signed": {
+    "_type": "timestamp",
+    "spec_version": "1.0.0",
+    "version": 1589415932,
+    "expires": "2020-05-19T00:25:32.891810Z",
```
I don't see `ExpirationEnforcement::Unsafe` anywhere, how does this work when we're past this date?

webern replied:
`ExpirationEnforcement::Unsafe` is a setting or the tough client, not the repository. The expiration dates in the repository are designed to ensure that we are using `ExpirationEnforcement::Unsafe` when we load it. The setting in question is [here](https://github.com/webern/bottlerocket/blob/0a50744bd6102b28eb6b5c74ecfd2d66687d6731/sources/api/migration/migrator/src/main.rs#L111) (on a branch that I won't be force pushing to).

tjkirch replied:
Ok - maybe set this to 1970-01-01 to make that more obvious?

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -387,7 +337,41 @@ impl Update {
+}
+
+// returns the versions with the lower version first.
+fn order_versions(a: &Version, b: &Version) -> (Version, Version) {
+    if let Some(direction) = Direction::from_versions(&a, &b) {
+        if direction == Direction::Backward {
+            return (b.clone(), a.clone());
+        }
+    }
+    (a.clone(), b.clone())
+}
```
Not sure I understand why we need `Direction` here, since `Version` is `Ord`... can we use the standard `Ordering`?  Something like this:
```rust
let lower, higher = match from.cmp(to) {
   Ordering::Less | Ordering::Equal => (from, to),
   Ordering::Greater => (to, from),
}
```

Is there a subtle difference I'm missing?  If so, we're using `Direction::from_versions` just before we call this function anyway, it seems like we could inline this conditional to simplify.

webern replied:
I made this simplification. I'm going to resolve conversations like these as I commit the fixes on my side branch (so that I know what I've addressed and what I haven't). I will aggregate several before doing a force push to the PR branch. You can wait for my re-review signal to avoid churn.

tjkirch commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -34,8 +36,11 @@ fn usage_msg<S: AsRef<str>>(msg: S) -> ! {
     pub(crate) migrate_to_version: Version,
+    pub(crate) root_path: PathBuf,
+    pub(crate) metadata_directory: PathBuf,
+    pub(crate) working_directory: PathBuf,
```
It feels really weird to make the user provide a temporary/working directory... can we use a tempdir instead?  If it has to be persistent, can we use another name for it?

webern replied:
A tempdir would be more convenient. Checking around a little bit to see if it's fine to actually use `tempfile::Tempdir` (i.e. are we fine with `/tmp/xxx/...`) or if we need to have a special place for these.

tjkirch replied:
Yeah - we should make an effort to clean it up before we quit (whether success or error) but they should eventually be cleaned up by the system in the worst case.

tjkirch replied:
Ah, I had only looked at the first force push when I made my last comment, and I see tempdir usage was added after.  I don't see anything removing them, though?  I think we need to make a good effort to remove them, whether overall we succeed or fail.

webern replied:
> Ah, I had only looked at the first force push when I made my last comment, and I see tempdir usage was added after. I don't see anything removing them, though? I think we need to make a good effort to remove them, whether overall we succeed or fail.

The beauty of `tempfile::TempDir` is that it uses the `drop` trait to do the delete. The only way they would not be cleaned up is if the program actually crashed, I think they even cleanup on panic.

tjkirch replied:
Ah, very nice, I hadn't seen the Drop implementation.

tjkirch commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -15,7 +15,9 @@ fn usage() -> ! {
         r"Usage: {}
             --datastore-path PATH
-            --migration-directories PATH[:PATH:PATH...]
+            --migration-directory PATH
+            --root-path PATH
+            --metadata-directory PATH
```
Need `--working-directory` in usage too, right?  (though honestly I'm hoping we can get rid of it, see other comment...)

webern replied:
I've added it temporarily pending its disappearance in favor of a tempdir.

tjkirch commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -114,15 +118,45 @@ impl Args {
 
         Self {
-            datastore_path: datastore_path.unwrap_or_else(|| usage()),
+            datastore_path: datastore_path.unwrap_or_else(|| usage_msg("datastore_path is empty")),
```
I think in the usage messages it's more helpful to talk in the terms the user is familiar with, e.g. `--datastore-path` rather than `datastore_path`.

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -41,10 +63,10 @@ pub(crate) enum Error {
-                    output.status.code()
-                        .map(|i| i.to_string()).unwrap_or_else(|| "signal".to_string()),
-                    std::str::from_utf8(&output.stderr)
-                        .unwrap_or_else(|_e| "<invalid UTF-8>")))]
+    output.status.code()
+    .map(| i | i.to_string()).unwrap_or_else(|| "signal".to_string()),
+    std::str::from_utf8(& output.stderr)
+    .unwrap_or_else(| _e | "<invalid UTF-8>")))]
```
rustfmt does this?!  I've never seen it insert spaces like that, in particular...

webern replied:
actually i think clion did that, i've disabled clion's formatting and restored this blob to the way it was. now i'll make sure only rustfmt runs on save. I also restored the newlines that clion ate in the subsequent error members.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -95,15 +124,28 @@ fn run() -> Result<()> {
+            &rundir,
         )?;
         flip_to_new_version(&args.migrate_to_version, &copy_path)?;
+        std::fs::remove_dir_all(args.working_directory.join(RUNDIR)).context(
```
```suggestion
        std::fs::remove_dir_all(&rundir).context(
```

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -95,15 +124,28 @@ fn run() -> Result<()> {
         flip_to_new_version(&args.migrate_to_version, &copy_path)?;
+        std::fs::remove_dir_all(args.working_directory.join(RUNDIR)).context(
+            error::DeleteDirectory {
+                path: args.working_directory.join(RUNDIR),
```
```suggestion
                path: &rundir,
```

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +494,204 @@ where
+/// Converts a filepath into a URI formatted string
+fn dir_url<P: AsRef<Path>>(path: P) -> Result<String> {
+    let url_result = Url::from_directory_path(&path);
+    // TODO - I can't figure out how to use .context with this error type
```
You said you were going to remove 'url' for file paths anyway, right?  If not, I imagine this code could be something like `let path = path.as_ref(); Url::from_directory_path(&path).unwrap_or_else(|| error::DirectoryUrl { path }.fail())` maybe?

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -349,14 +237,38 @@ where
+                migration: migration.to_owned(),
+            })?
+            .context(error::MigrationNotFound {
+                migration: migration.to_owned(),
```
```suggestion
                migration,
            })?
            .context(error::MigrationNotFound {
                migration,
```
You shouldn't need to convert to owned yourself, Snafu does that, and only as necessary.  (If you're doing a clone, or something else that needs a few cycles, you should aim for `.with_context(|| bla)` instead of a plain context, so it only happens in the failure case and not always.  But this case is so common that Snafu does it.)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,15 +215,16 @@ where
     direction: Direction,
-    migrations: &[P1],
-    source_datastore: P2,
+    migrations: &[String],
```
Does this have to take owned Strings?  It looks like we're only using them as references.  Would `&str` work?  Strings should automatically dereference to that.

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -59,18 +81,58 @@ pub(crate) enum Error {
     #[snafu(display("Failed reading migration directory entry: {}", source))]
     ReadMigrationEntry { source: io::Error },
-
+    #[snafu(display("Failed to load tough repo: {}", source))]
```
It's not letting me leave a suggestion here, but can we add a blank line separating these?

tjkirch replied:
Can't leave a suggestion on this line for some reason, but "tough" -> "TUF"

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +85,37 @@ fn run() -> Result<()> {
-        &current_version,
-        &args.migrate_to_version,
-    )?;
+    // Prepare to load the locally cached tough repository to obtain the manifest.
```
```suggestion
    // Prepare to load the locally cached TUF repository to obtain the manifest.
```

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +85,37 @@ fn run() -> Result<()> {
+    let migrations_url = dir_url(&args.migration_directory)?;
+
+    // Failure to load the tough repo at the expected location is a serious issue because updog
+    // should always create a tough repo that contains at least the manifest, even if there are no
```
```suggestion
    // Failure to load the TUF repo at the expected location is a serious issue because updog
    // should always create a TUF repo that contains at least the manifest, even if there are no
```

tjkirch commented on sources/updater/updog/src/main.rs:
```diff
@@ -32,6 +31,7 @@ const TARGET_ARCH: &str = "aarch64";
 const TRUSTED_ROOT_PATH: &str = "/usr/share/updog/root.json";
 const MIGRATION_PATH: &str = "/var/lib/bottlerocket-migrations";
 const METADATA_PATH: &str = "/var/cache/bottlerocket-metadata";
+const TOUGH_DATASTORE: &str = "/var/lib/bottlerocket/updog";
```
I still have no idea of the difference between these two things - can we please document them?  And explain why one is in our private partition (which is for small, secure, persistent artifacts, which have to be even more carefully managed for SELinux) and the other is temporary cache we don't need over a reboot?

webern replied:
For now I've added comments to explain what they are. `TOUGH_DATASTORE` should go away in favor of a tempdir. I'll add a TODO list to the PR description.

tjkirch commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -18,26 +18,18 @@ use std::fs;
 use std::path::Path;
-use std::str::FromStr;
 use tough;
+use tough::Limits;
```
```suggestion
use tough::{self, Limits};
```

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -30,37 +30,41 @@ use simplelog::{Config as LogConfig, TermLogger, TerminalMode};
     // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
-    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed)
-        .context(error::Logger)?;
+    if let Err(e) = TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed) {
+        info!("Term logger init returned an error: {}", e)
+    }
```
I'm not sure about this change.  I agree that migration is critical to startup, but I think I'd prefer if we fall back to stdout/stderr logging through some other method, or if we still have migrator fail but make the system reboot into the old version when it happens.

At a minimum I think this is unrelated to the main thrust of the change and I'd ask that it be done separately.

webern replied:
I needed to do this since calling init more than once panics (spent a lot of time figuring out why my tests were panicking, but not when they ran one at a time). I could move the init outside of the run function and place it in `main` instead as an alternative fix. That's what I'll do for now.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +85,37 @@ fn run() -> Result<()> {
+    // should always create a tough repo that contains at least the manifest, even if there are no
+    // migrations.
+    let repo = tough::Repository::load(
+        &tough::FilesystemTransport {},
```
```suggestion
        &tough::FilesystemTransport,
```
Don't need the curlies for a struct with no fields.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +85,37 @@ fn run() -> Result<()> {
+            datastore: &tough_datastore,
+            metadata_base_url: metadata_url.as_str(),
+            targets_base_url: migrations_url.as_str(),
+            limits: Default::default(),
```
Using default here but specifying values in update_metadata feels like a potential issue because if they don't match or they're changed in one place we could wind up with a repo that gets downloaded but won't be loaded here.  I think we should just use default in both.

webern replied:
Good catch, I had specifically created these settings in the shared `update_metadata` and forgot to use them here. The author of `updog` had spelled the limits out instead of using `default()`, so the version I'll push next is what I intended, both `updog` and `migrator` getting them from the same hard-coded place.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -95,15 +124,28 @@ fn run() -> Result<()> {
         )?;
         flip_to_new_version(&args.migrate_to_version, &copy_path)?;
+        std::fs::remove_dir_all(args.working_directory.join(RUNDIR)).context(
+            error::DeleteDirectory {
+                path: args.working_directory.join(RUNDIR),
+            },
+        )?;
     }
-
+    std::fs::remove_dir_all(&tough_datastore).context(error::DeleteDirectory {
+        path: tough_datastore,
+    })?;
```
I'm not sure it's a good idea to add these failures cases after the `flip_to_new_version` call.  If that call succeeds, we *are* on the new version, and signaling otherwise with the return code could cause rollbacks (or worse) when in fact we just have a straggler directory.  It feels like we should just warn if they fail to delete?

webern replied:
Excellent point, changing as suggested.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,15 +215,16 @@ where
+    migrations: &[String],
+    source_datastore: P,
     new_version: &Version,
+    migrations_rundir: &PathBuf,
```
Normally a reference to the owned type can instead be the reference type - it works with Rust's `Deref`.  This can be `&Path` and basically do the same things but not require the user have an owned `PathBuf`.  (It's sometimes better to go a step further and make it `AsRef<Path>` so it's more compatible whether the caller has a `Path` or a `PathBuf`.)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -349,14 +237,38 @@ where
+        {
+            let mut f = OpenOptions::new()
+                .write(true)
+                .create(true)
```
```suggestion
                .create(true)
                .truncate(true)
```
(I know we create the directory, but a bit of safety...)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -349,14 +237,38 @@ where
+                .create(true)
+                .open(&exec_path)
+                .context(error::MigrationSave {
+                    path: exec_path.clone(),
```
```suggestion
                    path: &exec_path,
```
Should be able to let Snafu handle that, and only clone on error cases.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +494,204 @@ where
-            vec![m02_1]
-        );
+    pub fn test_data() -> PathBuf {
+        let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
```
I don't see any of the test data being cleaned up - I think we want to remove this directory on test success, and remove it when starting the tests if it exists, and add it to gitignore...

tjkirch replied:
Sorry, I didn't see that the symlinks and things are being created in the `info.tmp` tempdir.

webern commented:
[push](https://github.com/bottlerocket-os/bottlerocket/compare/0a50744bd6102b28eb6b5c74ecfd2d66687d6731..3b57a357c1a3d9542d36ad22a0cd657b29a33db7) to 3b57a35 addresses most of the comments, with the exception of using a tempdir in updog and migrator.

webern commented:
[push](https://github.com/bottlerocket-os/bottlerocket/compare/3b57a357c1a3d9542d36ad22a0cd657b29a33db7..3f627811556fdd869edc856f14cb52f3905c4b7d) to 3f62781 uses `tempfile::TempDIr` in both `updog` and `migrator` for the `tough` datastore. Tested with the full test cycle.

webern commented on sources/api/migration/migrator/Cargo.toml:
```diff
@@ -18,10 +18,16 @@ semver = "0.9"
 cargo-readme = "3.1"
 
+[dev-dependencies]
+tempfile = "3.1"
```
remove `tempfile` from `dev-dependencies` as now it's also in the `dependencies` list.

webern commented on sources/updater/updog/Cargo.toml:
```diff
@@ -23,6 +22,7 @@ simplelog = "0.7"
 snafu = "0.6.0"
 toml = "0.5.1"
 tough = { version = "0.5.0", features = ["http"] }
+tempfile = "3.1.0"
```
move `tempdir` above `toml` (alphabetical ordering)

tjkirch commented on sources/api/migration/migrator/tests/data/repository/targets/a-second-migration.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Can we compress these as part of the testing rather than including binary artifacts?  It's weird to have the binaries, and it's likely someone will forget to change the binary if they change the script.

webern replied:
Not without signing the TUF repo on the fly, which I plan to do once we do a release of the `RepositoryEditor` from @zmrow.

zmrow replied:
> Not without signing the TUF repo on the fly, which I plan to do once we do a release of the `RepositoryEditor` from @zmrow.

Coming soon! :)

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/api/migration/migrator/tests/data/repository/targets/a-second-migration.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -eo pipefail
+migration_name="a-second-migration"
```
It seems the only difference between the scripts is this name, and I think it can be `${0##*/}`, and then they're the same script, maybe just symlinked?

webern replied:
> It seems the only difference between the scripts is this name, and I think it can be `${0##*/}`, and then they're the same script, maybe just symlinked?

Makes sense, I'll add it to the TODO list for test cleanup and do this when I generate the TUF repo at runtime.

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -114,15 +117,36 @@ impl Args {
         Self {
-            datastore_path: datastore_path.unwrap_or_else(|| usage()),
+            datastore_path: datastore_path
+                .unwrap_or_else(|| usage_msg("--datastore_path is empty")),
```
```suggestion
                .unwrap_or_else(|| usage_msg("--datastore-path must be specified")),
```
Kind of sounded like it was specified but given an empty argument.

(Note also the underscore -> hyphen)

Same below.

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -59,9 +80,58 @@ pub(crate) enum Error {
+    },
+
+    #[snafu(display("Unable to create URL from path '{}'", path.display()))]
+    PathUrl { path: PathBuf, backtrace: Backtrace },
```
Hmm, I'm not sure this represents the failure case - it's not related to URLs exactly, it'd be if a given path was not valid UTF-8.

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +476,193 @@ where
+        let lines: Vec<&str> = contents.split('\n').collect();
+        assert_eq!(lines.len(), 3);
+        let first_line = *lines.get(0).unwrap();
+        if !first_line.contains("x-first-migration: --forward") {
+            panic!(format!(
```
Can we simplify these checks using `assert!` instead of `if` and `panic`?

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch commented on sources/api/migration/migrator/tests/data/repository/metadata/1.root.json:
```diff
@@ -0,0 +1,71 @@
+      }
+    },
+    "version": 1,
+    "keys": {
```
Out of curiosity, what was the change to the keys in the force push?

webern replied:
> Out of curiosity, what was the change to the keys in the force push?

The keys are throw-away, I'm not saving them somewhere, I have a script that will regenerate the TUF repo whenever I need to make a change, and that includes generating new keys. This problem will go away once the TUF repo is created during the test.

tjkirch replied:
Ah, cool.  And we can do that now with https://github.com/awslabs/tough/pull/125/ merged?

[edit] Ah, after the upcoming `tough` release.

webern replied:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

tjkirch reviewed (commented):
Hoping @iliana can confirm the tempdir for tough datastore is OK.

tjkirch commented:
@webern confirmed for me that this force-push ^ was just a rebase.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 649b11a to ee995a6 now

This adds backward compatibility with unsigned migrations.

iliana commented:
> Hoping @iliana can confirm the tempdir for tough datastore is OK.

I know @webern and I discussed it, and I was confident that the tempdir was fine for this particular case. I'll try and write up some full details, and maybe those can turn into a doc comment on tough. :)

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from ee995a6 to f481a2d 3 minutes ago

This is a rebase from develop that bumps the version of `tough`.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441

This push addresses:

  * Use `tough::RepoEditor` to create the migrator unit test repo on the fly.
  * `migrator/src/args.rs` review comment (`unwrap_or_else` messages).
  * `migrator/src/error.rs` review comment (`PathUrl` unclear).

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 4443441 to dab3f59 41 seconds ago

Small: cleanup in `migrator/srv/error.rs`

iliana commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -23,45 +23,89 @@
-use error::Result;
+lazy_static! {
+    /// This is the last version of Bottlerocket that supports *only* unsigned migrations.
+    static ref LAST_UNSIGNED_MIGRATIONS_VERSION: Version = Version::new(0, 3, 4);
```
:thinking: I feel like `Version::new` could be a `const fn`...

webern replied:
Unfortunately the compiler yelled at me because there is a Vec or a String inside (because they support suffixes like "alpha"). So allocation occurs. If it's too distasteful, I could just inline the ::new or return the Version from a function.

iliana replied:
Oh, right, this is `semver::Version`. See also https://github.com/steveklabnik/semver/issues/200#issuecomment-646316701

iliana commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +125,54 @@ fn run() -> Result<()> {
+    // DEPRECATED CODE END /////////////////////////////////////////////////////////////////////////
+
+    // Prepare to load the locally cached TUF repository to obtain the manifest.
+    let tough_datastore = TempDir::new().context(error::CreateToughTempDir)?;
```
A bit more on the topic of using a TempDir for the datastore here, since @tjkirch asked:

The threats TUF mitigates are more than the threats we are attempting to mitigate here by caching signatures for migrations locally and using them after a reboot but prior to internet connectivity. We are caching the TUF repo and use it while offline after a reboot to mitigate binaries being added or modified in the migrations directory; the TUF repo is simply a code signing method we already have in place, even if it's not one that initially makes sense for this use case.

Part of doing this is disabling timestamp checking, because we want an instance to still come up and run migrations if someone runs an update, then shuts down the instance for several weeks (beyond the expiration of at least the cached timestamp.json) before booting it back up again.

Similarly, we see no value in keeping a datastore around, because the latest known versions of the repository metadata will *always* be the versions of repository metadata we have cached on the disk.

bcressey replied:
Let's capture some of this explanation in a comment.

webern replied:
Addressed in push to [f5a28e5](https://github.com/bottlerocket-os/bottlerocket/compare/50cb08df7a016189eae37a20449b1779d9786331..f5a28e55ec6f9ebfde8b8e720b0aee0689dee61b)

iliana commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,8 +422,131 @@ where
+        if !output.stdout.is_empty() {
+            debug!(
+                "Migration stdout: {}",
+                std::str::from_utf8(&output.stdout).unwrap_or("<invalid UTF-8>")
```
Losing the migration binary's output just because it isn't valid UTF-8 doesn't seem like a good idea. What about using [`String::from_utf8_lossy`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy)?

webern replied:
Addressed in push to [00ce955](https://github.com/bottlerocket-os/bottlerocket/compare/f5a28e55ec6f9ebfde8b8e720b0aee0689dee61b..00ce9553e6dfcb1dd09b35361e478cd29502f2bb)

iliana commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,8 +422,131 @@ where
+        // TODO - remove this use of the filesystem when we add pentacle
+        let exec_path = migrations_rundir.as_ref().join(&migration);
+        {
+            let mut f = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .truncate(true)
+                .open(&exec_path)
```
This is exactly [`File::create`](https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.create).

webern replied:
It went away with pentacle anyway.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from dab3f59 to b3a4354 14 seconds ago

Small cleanups.

iliana commented on sources/api/migration/migrator/tests/data/fake-key.pem:
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
We call this `snakeoil.pem` [in tough](https://github.com/awslabs/tough/blob/develop/tough/tests/data/snakeoil.pem), might be nice to do the same here so it's easier to identify keys made for test purposes.

webern replied:
Addressed in push to [00ce955](https://github.com/bottlerocket-os/bottlerocket/compare/f5a28e55ec6f9ebfde8b8e720b0aee0689dee61b..00ce9553e6dfcb1dd09b35361e478cd29502f2bb)

iliana commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,8 +419,131 @@ where
+        // deflate with an lz4 decoder read
+        let mut reader = lz4::Decoder::new(lz4_bytes).context(error::Lz4Decode { migration })?;
+
+        // TODO - remove this use of the filesystem when we add pentacle
+        let exec_path = migrations_rundir.as_ref().join(&migration);
+        {
+            let mut f = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .truncate(true)
+                .open(&exec_path)
+                .context(error::MigrationSave { path: &exec_path })?;
+            let _ = std::io::copy(&mut reader, &mut f)
+                .context(error::MigrationSave { path: &exec_path })?;
+        }
+
+        // Ensure the migration is executable.
+        fs::set_permissions(&exec_path, Permissions::from_mode(0o755))
+            .context(error::SetPermissions { path: &exec_path })?;
+
+        let mut command = Command::new(&exec_path);
```
Getting pentacle working here *should* be as trivial as:

```suggestion
        let mut command = pentacle::SealedCommand::new(&mut reader)
            .context(error::SealMigration)?;
```

along with appropriate changes in the error module. If this *doesn't* work, it might be a bug in pentacle. :)

webern replied:
Implemented in push to [50cb08d](https://github.com/bottlerocket-os/bottlerocket/compare/b3a435433390378699891a01566eff7e74305099..50cb08df7a016189eae37a20449b1779d9786331)

bcressey commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -15,7 +15,10 @@ fn usage() -> ! {
+            --migration-directory PATH
+            --root-path PATH
+            --metadata-directory PATH
+            --working-directory PATH
```
`working-directory` can go away again since it's no longer supported.

webern replied:
Addressed in push to [f5a28e5](https://github.com/bottlerocket-os/bottlerocket/compare/50cb08df7a016189eae37a20449b1779d9786331..f5a28e55ec6f9ebfde8b8e720b0aee0689dee61b)

bcressey commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -39,6 +48,14 @@ lazy_static! {
                    $").unwrap();
 }
 
+/// These are the limits that Bottlerocket will use for the `tough` library.
+pub const REPOSITORY_LIMITS: Limits = Limits {
+    max_root_size: 1024 * 1024,         // 1 MiB
+    max_targets_size: 1024 * 1024 * 10, // 10 MiB
+    max_timestamp_size: 1024 * 1024,    // 1 MiB
+    max_root_updates: 1024,
+};
```
We should have a way to track these sizes over time so we'll know if we're getting close to any of these limits. Maybe a `tuftool` feature to dump stats.

webern replied:
Added https://github.com/awslabs/tough/issues/156

bcressey commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -1,33 +1,42 @@
 
 pub const MAX_SEED: u32 = 2048;
 
+// the use of this regex is deprecated and only used for backward compatibility with
+// unsigned migration
```
nit: maybe add the same DEPRECATED CODE BEGIN/END markers around this part

webern replied:
Addressed in push to [f5a28e5](https://github.com/bottlerocket-os/bottlerocket/compare/50cb08df7a016189eae37a20449b1779d9786331..f5a28e55ec6f9ebfde8b8e720b0aee0689dee61b)

bcressey reviewed (approved):
LGTM

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from b3a4354 to 50cb08d

Added `pentacle` tested with the full Bottlerocket test cycle. I did some acrobatics to preserve my `migrator` end-to-end "unit" test, but it works.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 50cb08d to f5a28e5

Addresses Ben's comments.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from f5a28e5 to 00ce955

Addresses iliana's smaller comments.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 00ce955 to acd95cf 16 seconds ago

Fix a one word oops.

tjkirch commented on sources/api/migration/migrator/Cargo.toml:
```diff
@@ -12,16 +12,23 @@ bottlerocket-release = { path = "../../../bottlerocket-release" }
 snafu = "0.6"
 update_metadata = { path = "../../../updater/update_metadata" }
+tempfile = "3.1.0"
+tough = { version = "0.6.0" }
```
```suggestion
tough = "0.6.0"
```

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/args.rs:
```diff
@@ -114,15 +116,36 @@ impl Args {
+            migration_directory: migration_directory
+                .unwrap_or_else(|| usage_msg("--migration-directory must be specified")),
+            migrate_to_version: migrate_to_version
+                .unwrap_or_else(|| usage_msg("--migrate-to-version could not be determined")),
```
```suggestion
                .unwrap_or_else(|| usage_msg("Desired version could not be determined; pass --migrate-to-version or --migrate-to-version-from-os-release")),
```

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -13,6 +13,12 @@ pub(crate) enum Error {
     Internal { msg: String },
 
+    #[snafu(display("Unable to create tempdir for migration binaries: '{}'", source))]
+    CreateRunDir { source: std::io::Error },
```
Is this unused?

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -22,6 +28,21 @@ pub(crate) enum Error {
+    },
+
+    #[snafu(display("Failed to convert '{}' to a URL", path.display()))]
+    DirectoryUrl { path: PathBuf, backtrace: Backtrace },
```
Are these unused?  I'm not sure if GitHub is barfing and not showing me everything...

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -59,9 +83,55 @@ pub(crate) enum Error {
+    #[snafu(display("Error loading manifest: {}", source))]
+    LoadManifest {
+        source: update_metadata::error::Error,
+        backtrace: Backtrace,
```
Is there a methodology for when you're bringing in Backtrace and when you're not?  I don't personally think they're very helpful, so I'd vote for leaving it out unless there's some methodology.

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/error.rs:
```diff
@@ -59,9 +83,55 @@ pub(crate) enum Error {
+    },
+
+    #[snafu(display("Error saving migration '{}': {}", path.display(), source))]
+    MigrationSave {
```
Unused?

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/tests/data/migration.rs:
```diff
@@ -0,0 +1,47 @@
+use std::str::FromStr;
+
+/// This program serves as a pseudo-migration for testing. It will write its name and arguments to
+/// a file named `results.txt` one level above the directory it receives as the `--source-datastore`
```
```suggestion
/// a file named `result.txt` one level above the directory it receives as the `--source-datastore`
```

webern replied:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

tjkirch commented on sources/api/migration/migrator/tests/data/migration.rs:
```diff
@@ -0,0 +1,47 @@
+    assert!(!outdir.to_string_lossy().is_empty());
+    let outfile_path = outdir.join("result.txt");
+    // create or append
+    let mut f = if !outfile_path.is_file() {
```
I'm not sure I understand this conditional.  `create(true)` is the only difference, and it's `true` if the file doesn't exist, but you can also use `create(true)` when the file does exist and it'll just open the existing file.  Seems like we can use a single `OpenOptions`?  (I'm assuming the important part isn't checking whether it's a *file*, per se, because this is test output that we fully control and it shouldn't be a directory or anything else.)

webern replied:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

tjkirch commented on sources/api/migration/migrator/tests/data/migration.rs:
```diff
@@ -0,0 +1,47 @@
+                datastore_argument = arg.to_string();
+            }
+        }
+    }
```
```suggestion
    let args = args.next().expect("Test migration missing 0th argument..?!")
    datastore_argument = args.nth(2).expect("Test migration not passed enough arguments, got {}", args.join(" "));
    message.push_str(args.join(" "));
```
Untested, but I feel like it can be simplified a bit.

webern replied:
Didn't work on the first try, but...

> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

tjkirch commented on sources/updater/update_metadata/src/direction.rs:
```diff
@@ -2,13 +2,13 @@
 /// Direction represents whether we're moving forward toward a newer version, or rolling back to
 /// an older version.
 #[derive(Debug, Copy, Clone, PartialEq, Eq)]
-pub(crate) enum Direction {
```
Looks like we're not using Direction outside of migrator, can we move it back and avoid some of these changes?

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

Though it looks like rustfmt made some changes to the formatting.

tjkirch commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -1,33 +1,43 @@
+use tough::{self, Limits};
+
+#[macro_use]
+extern crate lazy_static;
```
```suggestion
use lazy_static::lazy_static;
```

This syntax is deprecated, can we stick with the previous `use lazy_static::lazy_static;` statement up above?

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/updater/update_metadata/src/lib.rs:
```diff
@@ -38,6 +48,15 @@ lazy_static! {
+// DEPRECATED CODE END /////////////////////////////////////////////////////////////////////////////
+
+/// These are the limits that Bottlerocket will use for the `tough` library.
+pub const REPOSITORY_LIMITS: Limits = Limits {
```
Can we stick with the `Defaults` value for `Limits` instead?  We've defined these in a number of places, which worries me a little, and I don't think we have a reason to deviate.

webern replied:
> Can we stick with the `Defaults` value for `Limits` instead? We've defined these in a number of places, which worries me a little, and I don't think we have a reason to deviate.

It's only in one place now, but I'm happy to revert to default(), no problem.

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +783,300 @@ where
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
 #[cfg(test)]
 mod test {
```
There's a tremendous amount of code for testing here - can we move this module to its own file?

webern replied:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

tjkirch commented on sources/updater/updog/src/main.rs:
```diff
@@ -210,10 +214,17 @@ fn retrieve_migrations(
+    let mut targets = find_migrations(start, target, &manifest)?;
+
+    // DEPRECATED CODE BEGIN ///////////////////////////////////////////////////////////////////////
+    // write unsigned migrations for backward compatibility. note that signed migrations will have
```
It might also be worth a note in here about how the compat migrations are deflated before writing, where the signed migrations are not, so they can be validated as they were in the repo?

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -23,45 +23,88 @@
+use tough::ExpirationEnforcement;
+use update_metadata::{load_manifest, Direction, MIGRATION_FILENAME_RE, REPOSITORY_LIMITS};
+
+#[macro_use]
+extern crate lazy_static;
```
```suggestion
use lazy_static::lazy_static;
```

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -23,45 +23,88 @@
 
-fn run() -> Result<()> {
-    let args = Args::from_env(env::args());
+#[deprecated(since = "0.3.5", note = "for unsigned migrations.")]
```
```suggestion
```
It doesn't seem like this function is deprecated?  You'd still use it to determine whether migrations are signed after that version...

webern replied:
Addressed in [e864eba](https://github.com/bottlerocket-os/bottlerocket/compare/acd95cfacb27e120daf9e903dd4b6cd0f2f8a015..e864eba052e35a165bd07ae9a04c1bb410daf6bd)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+    // DEPRECATED CODE BEGIN ///////////////////////////////////////////////////////////////////////
+    // check if the `from_version` supports signed migrations. if not, run the 'old'
+    // unsigned migrations code and return.
+    if !are_migrations_signed(&current_version) {
```
```suggestion
```
I'd rather remove the code entirely when we don't want to support unsigned migrations.  I've found it's usually not worth having the branches, because they can go wrong on their own, you can forget them, etc.

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -281,16 +348,47 @@ fn select_migrations<P: AsRef<Path>>(
+    select_unsigned_migrations(from, to, &candidates)
+}
+
+fn get_current_version<P>(datastore_dir: P) -> Result<Version>
```
It looks like this function was just moved without change?  Can we put it back, to reduce the review effort in the PR and in future investigations?

webern replied:
> It looks like this function was just moved without change? Can we put it back, to reduce the review effort in the PR and in future investigations?

Maybe? It looks like line 106 is available. Hopefully if I paste it there then git won't show it as having been moved. We'll see.

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+    // check if the `from_version` supports signed migrations. if not, run the 'old'
+    // unsigned migrations code and return.
+    if !are_migrations_signed(&current_version) {
+        // note in the system journal that the unsigned code path ran.
+        eprintln!("migrator is running unsigned migrations");
+        warn!("migrator is running unsigned migrations");
+        return find_and_run_unsigned_migrations(
+            &args.migration_directory,
+            &args.datastore_path, // TODO(brigmatt) make sure this is correct
+            &current_version,
+            &args.migrate_to_version,
+            &direction,
+        );
+    } else {
+        // note in the system journal that the signed code path ran.
+        info!("migrator is running signed migrations");
+        eprintln!("migrator is running signed migrations");
```
I don't think we need these messages in the logs/journal, we know which type of migrations ran based on the version.  (Ties with my other comments about simplifying/removing the branching.)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+    // versions of the repository metadata will always be the versions of repository metadata we
+    // have cached on the disk. More info at `ExpirationEnforcement::Unsafe` below.
+    let tough_datastore = TempDir::new().context(error::CreateToughTempDir)?;
+    let metadata_url = dir_url(&args.metadata_directory)?;
+    let migrations_url = dir_url(&args.migration_directory)?;
```
I think we can use https://docs.rs/url/2.1.1/url/struct.Url.html#method.from_directory_path instead of our own implementation.

webern replied:
Addressed in [1f69203](https://github.com/bottlerocket-os/bottlerocket/compare/e864eba052e35a165bd07ae9a04c1bb410daf6bd..1f69203f6dc37958505c0a92c63daf6245f59b3b)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+            // prior to Internet connectivity. We are caching the TUF repo and use it while offline
+            // after a reboot to mitigate binaries being added or modified in the migrations
+            // directory; the TUF repo is simply a code signing method we already have in place,
+            // even if it's not one that initially makes sense for this use case.
+            expiration_enforcement: ExpirationEnforcement::Unsafe,
```
```suggestion
            // even if it's not one that initially makes sense for this use case.
            // So, we don't care if the targets expired between updog downloading them and now.
            expiration_enforcement: ExpirationEnforcement::Unsafe,
```
I feel like the explanation was good but didn't actually say why / that we don't care about expiration :)

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,8 +426,112 @@ where
+        // deflate with an lz4 decoder read
+        let mut reader = lz4::Decoder::new(lz4_bytes).context(error::Lz4Decode { migration })?;
+
+        // create a sealed command with pentacle
```
```suggestion
        // Create a sealed command with pentacle, so we can run the verified bytes from memory
```

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -328,8 +426,112 @@ where
+            .context(error::LoadMigration { migration })?
+            .context(error::MigrationNotFound { migration })?;
+
+        // deflate with an lz4 decoder read
```
```suggestion
        // Add an LZ4 decoder so the bytes will be deflated on read
```

webern replied:
Addressed in [4f34544](https://github.com/bottlerocket-os/bottlerocket/compare/afc4b5d9ddfb7c6fb24f30abdaf6a5a000df222c..4f345446016e314c3da3f0855c92db5ca979f2a4)

tjkirch commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -580,60 +783,300 @@ where
+        let tempdir = TempDir::new().unwrap();
+        let source_file = tempdir.path().join("migration.rs");
+        std::fs::write(&source_file, sourcecode.as_bytes()).unwrap();
+        let compile_command_result = Command::new("rustc")
```
I have strong reservations about calling rustc in a test; it doesn't seem expected, reliable, performant, or safe in general.  My preferences would be (1) shell script with pentacle, which we have ideas on making work; (2) check in a binary (made properly small); (3) no test; (4) rustc.

webern replied:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

tjkirch replied:
We have the ability to run shell scripts with pentacle 0.2, which we can use to add the test back.  I think you wanted to do that in a separate PR; resolving.

webern commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+    // DEPRECATED CODE BEGIN ///////////////////////////////////////////////////////////////////////
+    // check if the `from_version` supports signed migrations. if not, run the 'old'
+    // unsigned migrations code and return.
+    if !are_migrations_signed(&current_version) {
```
@tjkirch left a comment here that, inexplicably, GitHub allowed me to delete when I was, in fact, trying to delete my own response. Fortunately I have both pasted to a text file:

@tjkirch wrote
> I'd rather remove the code entirely when we don't want to support unsigned migrations. I've found it's usually not worth having the branches, because they can go wrong on their own, you can forget them, etc.

And my response:
The branch is required for correctness. I don't understand what you're suggesting. If we don't check the version and choose a branch then we are either supporting signed or unsigned, but not both.

tjkirch replied:
Sorry, I guess that was a more general comment since I was looking at the branching/comments in multiple places regarding deprecation, and it was throwing me off overall.  Yeah, obviously we need to pick here, but we have more pertinent/correct information like the repo / signed migrations being available, right?  If that's too complicated I don't mind this.

webern replied:
We will stick with the current implementation. The main risk is that we need to change the value of the version constant if our release plans change. Lets both call it out if that happens! Thanks.

webern commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -82,11 +124,65 @@ fn run() -> Result<()> {
+    // unsigned migrations code and return.
+    if !are_migrations_signed(&current_version) {
+        // note in the system journal that the unsigned code path ran.
+        eprintln!("migrator is running unsigned migrations");
```
@tjkirch left a comment here that, inexplicably, GitHub allowed me to delete when I was, in fact, trying to delete my own response. Fortunately I have both pasted to a text file:

@tjkirch wrote:
> I don't think we need these messages in the logs/journal, we know which type of migrations ran based on the version. (Ties with my other comments about simplifying/removing the branching.)

And I replied:
I am using them during test cycles. I can delete them, but once I do I will no longer be able to verify correctness during the test cycles as I have been (see the test procedure in the PR comments).

tjkirch replied:
I'd prefer to use a more structured method like running different migrations on each path.  I'm hoping we can go back to the shell script method so having test migrations is trivial...

webern replied:
I meant when I run Bottlerocket and check the logs to ensure that migrator did what I want. However, I can remove these print statements as the last push.

tjkirch replied:
Yeah, I was suggesting that we use different test migrations (which can print different output) to get the same effect, and even check it in the test, rather than impacting the actual codebase here.

webern replied:
removed these in this push https://github.com/bottlerocket-os/bottlerocket/compare/b6fe3cd824df6df149da51efc14b01b222f46787..7b6b6c27a0abbeedd77535ba948aa0c7d220bb1c

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from acd95cf to e864eba 1 minute ago

Quickfixes for some of @tjkirch's comments. Not done yet...

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from e864eba to 1f69203 now

Uses `url::Url` instead of `format!("file:/{}", path)`

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from afc4b5d to 4f34544 now

Addresses most of @tjkirch's remaining comments.

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from 4f34544 to b6fe3cd now

`git rebase develop`

webern commented on sources/api/migration/migrator/src/main.rs:
```diff
@@ -579,61 +785,3 @@ where
-
-// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
-
-#[cfg(test)]
```
This can be restored, the function is select_unsigned_migrations

webern replied:
restored

tjkirch reviewed (approved):
No more blockers from me, as long as testing is still good.

webern commented:
Just did a successful test cycle with b6fe3cd

webern commented:
> webern force-pushed the webern:pr/signed-migrations branch from b6fe3cd to 7b6b6c2 1 minute ago

Last cleanup: restores the original unit test (which will be deleted once unsigned migrations are gone), and stops printing to the system journal per @tjkirch's request.

It's now ready to merge.
